Simplify L1 lookups for DSN farmers.

shamil-gadelshin opened the pull request:
This PR simplifies L1 piece lookups (from archival storage) for DSN farmers. Initially, we used cuckoo filters to request only farmers that highly likely contains the desired pieces. However, we had a downside of updating relatively large cuckoo filters on each sector replotting. After researching the current state-of-art algorithms like scalable bloom filter, morton filter, vector quotient filter, xor-filter, and others we decided to simplify the algorithm and request all connected farmers unconditionally. 

Changes:
- remove cuckoo filter from peer-info protocol,
- remove cuckoo filter from farmer and use connected peers instead.

Fixes https://github.com/subspace/subspace/issues/1627

### Code contributor checklist:
* [x] I have read, understood and followed [contributing guide](https://github.com/subspace/subspace/blob/main/CONTRIBUTING.md)

shamil-gadelshin committed:
Simplify L1 lookups.

- remove cuckoo filter from peer-info protocol.
- remove cuckoo filter form farmer and use connected peers instead

nazar-pc commented on crates/subspace-farmer/src/utils/readers_and_pieces.rs:
```diff
@@ -95,10 +92,6 @@ impl ReadersAndPieces {
             }
         }
-
-        if !new_piece_indices.is_empty() {
```
`new_piece_indices` needs to be removed now, it is no longer used

nazar-pc commented on crates/subspace-farmer/src/utils/readers_and_pieces.rs:
```diff
@@ -133,11 +126,6 @@ impl ReadersAndPieces {
             }
         }
-
-        if !deleted_piece_indices.is_empty() {
```
Same here, `deleted_piece_indices` should be deleted

nazar-pc commented on crates/subspace-networking/src/peer_info.rs:
```diff
@@ -28,32 +27,11 @@ pub struct Notification;
-        /// Peer info data.
-        cuckoo_filter: Arc<CuckooFilterDTO>,
-    },
+    Farmer,
```
Isn't this a breaking change on the network level?

shamil-gadelshin replied:
Yes, it is. I added a related label.

nazar-pc commented on crates/subspace-networking/src/peer_info.rs:
```diff
@@ -143,19 +113,13 @@ impl PeerInfoProvider {
-            PeerInfoProvider::Farmer(provider) => provider.on_notification(handler),
-        }
+    pub fn on_notification(&self, _handler: NotificationHandler) -> Option<HandlerId> {
+        // This is a stub method for future usages of notification.
```
What notifications do you have in mind? If nothing specific I'd remove it.

shamil-gadelshin committed:
Apply review suggestions.

- remove obsolete code from `readers_and_pieces.rs`
- remove notifications from `peer-info` protocol
- fix `cargo doc` issue

shamil-gadelshin committed:
Merge branch 'main' into optimize-l1-requests

nazar-pc reviewed (changes requested):
PR label is not enough. If we are merging this into `main` and intend to release as Gemini 3f it needs to be compatible with Gemini 3f, hence it must be able to encode/decode data the way it was before.

Please restore non-unit `Farmer` variant doesn't need to be named `cuckoo_filter`, doesn't need to be wrapped in an `Arc`, doesn't even need to have one field as before, just encode/decode the same way. We need a a vector of bytes + `u64`. `TODO` should be added to remove enum fields in the future when we can introduce breaking changes.

If necessary read SCALE codec docs to understand how encoding/decoding works there.

shamil-gadelshin committed:
networking: Add backward compatibility for peer-info protocol.
